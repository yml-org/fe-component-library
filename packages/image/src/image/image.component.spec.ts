

import { LitElement } from 'lit-element';
import {
  imageConfig
} from '../constants/image.variant';

describe('image-component', () => {
  const IMAGE_COMPONENT = 'image-component';
  let imageElement: LitElement;

  const getShadowRoot = (tagName: string): ShadowRoot =>
    document.body.getElementsByTagName(tagName)[0].shadowRoot;

  beforeEach(() => {
    imageElement = window.document.createElement(IMAGE_COMPONENT) as LitElement;
    document.body.appendChild(imageElement);
  });

  afterEach(() => {
    document.body.getElementsByTagName(IMAGE_COMPONENT)[0].remove();
  });

  it('matches component snapshot', async () => {
    expect(
      document.body.getElementsByTagName(IMAGE_COMPONENT)[0]
    ).toMatchSnapshot();
  });

  it('sets part property correctly', async () => {
    const elem = getShadowRoot(IMAGE_COMPONENT).querySelector('img');
    expect(elem.part[0]).toBe("custom-image");
  });

  it('sets image alt text as secondary', async () => {
    imageElement['altText']=imageConfig.altText;
    window.document.body.appendChild(imageElement);
    await imageElement.updateComplete;
    const imageEle = getShadowRoot(IMAGE_COMPONENT).querySelector('img'); 
    expect(imageEle.alt).toBe(imageConfig.altText);
  });

  it('sets image width property', async () => {
    imageElement['size']=imageConfig.size;
    window.document.body.appendChild(imageElement);
    await imageElement.updateComplete;
    const imageEle = getShadowRoot(IMAGE_COMPONENT).querySelector('img'); 
    expect(imageEle.getAttribute('width')).toBe(imageConfig.size.width);
  });

  it('sets image height property', async () => {
    imageElement['size']=imageConfig.size;
    window.document.body.appendChild(imageElement);
    await imageElement.updateComplete;
    const imageEle = getShadowRoot(IMAGE_COMPONENT).querySelector('img'); 
    expect(imageEle.getAttribute('height')).toBe(imageConfig.size.height);
  });

  it('sets image max-height property', async () => {
    imageElement['size']=imageConfig.size;
    window.document.body.appendChild(imageElement);
    await imageElement.updateComplete;
    const imageEle = getShadowRoot(IMAGE_COMPONENT).querySelector('img'); 
    expect(imageEle.getAttribute('max-height')).toBe(imageConfig.size.maxHeight);
  });

  it('sets image max-width property', async () => {
    imageElement['size']=imageConfig.size;
    window.document.body.appendChild(imageElement);
    await imageElement.updateComplete;
    const imageEle = getShadowRoot(IMAGE_COMPONENT).querySelector('img'); 
    expect(imageEle.getAttribute('max-width')).toBe(imageConfig.size.maxWidth);
  });

  it('sets image min-width property', async () => {
    imageElement['size']=imageConfig.size;
    window.document.body.appendChild(imageElement);
    await imageElement.updateComplete;
    const imageEle = getShadowRoot(IMAGE_COMPONENT).querySelector('img'); 
    expect(imageEle.getAttribute('min-width')).toBe(imageConfig.size.minWidth);
  });

  it('sets image min-height property', async () => {
    imageElement['size']=imageConfig.size;
    window.document.body.appendChild(imageElement);
    await imageElement.updateComplete;
    const imageEle = getShadowRoot(IMAGE_COMPONENT).querySelector('img'); 
    expect(imageEle.getAttribute('min-height')).toBe(imageConfig.size.minHeight);
  });

  it('sets image object-fit property', async () => {
    imageElement['size']=imageConfig.size;
    window.document.body.appendChild(imageElement);
    await imageElement.updateComplete;
    const imageEle = getShadowRoot(IMAGE_COMPONENT).querySelector('img'); 
    expect(imageEle.classList.contains('object-fill')).toBe(true);
  });

  it('Does not set image object-fit property by default', async () => {
    const imageEle = getShadowRoot(IMAGE_COMPONENT).querySelector('img'); 
    expect(imageEle.classList.contains('object-fill')).toBe(false);
  });

  it('sets image rounded property to true if rounded=true', async () => {
    imageElement['size']=imageConfig.size;
    window.document.body.appendChild(imageElement);
    await imageElement.updateComplete;
    const imageEle = getShadowRoot(IMAGE_COMPONENT).querySelector('img'); 
    expect(imageEle.classList.contains('rounded-full')).toBe(true);
  });

  it('sets image rounded property to false by default', async () => {
    const imageEle = getShadowRoot(IMAGE_COMPONENT).querySelector('img'); 
    expect(imageEle.classList.contains('rounded-full')).toBe(false);
  });
  
  it('sets image border property', async () => {
    imageElement['size']=imageConfig.size;
    window.document.body.appendChild(imageElement);
    await imageElement.updateComplete;
    const imageEle = getShadowRoot(IMAGE_COMPONENT).querySelector('img'); 
    expect(imageEle.style.border).toContain(imageConfig.size.border.width + ' ' + imageConfig.size.border.type );
  });

  it('sets image src property', async () => {
    imageElement['placeholderImg']=imageConfig.placeholderImg;
    window.document.body.appendChild(imageElement);
    await imageElement.updateComplete;
    const imageEle = getShadowRoot(IMAGE_COMPONENT).querySelector('img'); 
    expect(imageEle.getAttribute('src')).toBe(imageConfig.placeholderImg);
  });

  it('sets image srcset property', async () => {
    imageElement['src']=imageConfig.src;
    window.document.body.appendChild(imageElement);
    await imageElement.updateComplete;
    const imageEle = getShadowRoot(IMAGE_COMPONENT).querySelector('img'); 
    expect(imageEle.getAttribute('srcset')).toBe(imageConfig.src);
  });
});