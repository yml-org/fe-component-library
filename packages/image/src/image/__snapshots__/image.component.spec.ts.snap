// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`image-component matches component snapshot 1`] = `
N {
  "_$AL": Map {},
  "_$Do": W {
    "_$AA": <!---->,
    "_$AB": null,
    "_$AH": Ee {
      "_$AD": B {
        "el": <template />,
        "parts": Array [
          Object {
            "ctor": [Function],
            "index": 0,
            "name": "class",
            "strings": Array [
              "max-w-full h-auto transition-shadow ease-in-out duration-300
     ",
              "
     ",
              "",
            ],
            "type": 1,
          },
          Object {
            "ctor": [Function],
            "index": 0,
            "name": "src",
            "strings": Array [
              "",
              "",
            ],
            "type": 1,
          },
          Object {
            "ctor": [Function],
            "index": 0,
            "name": "srcset",
            "strings": Array [
              "",
              "",
            ],
            "type": 1,
          },
          Object {
            "ctor": [Function],
            "index": 0,
            "name": "alt",
            "strings": Array [
              "",
              "",
            ],
            "type": 1,
          },
          Object {
            "ctor": [Function],
            "index": 0,
            "name": "width",
            "strings": Array [
              "",
              "",
            ],
            "type": 1,
          },
          Object {
            "ctor": [Function],
            "index": 0,
            "name": "height",
            "strings": Array [
              "",
              "",
            ],
            "type": 1,
          },
          Object {
            "ctor": [Function],
            "index": 0,
            "name": "min-width",
            "strings": Array [
              "",
              "",
            ],
            "type": 1,
          },
          Object {
            "ctor": [Function],
            "index": 0,
            "name": "max-width",
            "strings": Array [
              "",
              "",
            ],
            "type": 1,
          },
          Object {
            "ctor": [Function],
            "index": 0,
            "name": "min-height",
            "strings": Array [
              "",
              "",
            ],
            "type": 1,
          },
          Object {
            "ctor": [Function],
            "index": 0,
            "name": "max-height",
            "strings": Array [
              "",
              "",
            ],
            "type": 1,
          },
          Object {
            "ctor": [Function],
            "index": 0,
            "name": "style",
            "strings": Array [
              "border:",
              "",
            ],
            "type": 1,
          },
        ],
      },
      "_$AM": [Circular],
      "_$AN": undefined,
      "u": Array [
        tt {
          "_$AH": Array [
            undefined,
            undefined,
          ],
          "_$AM": [Circular],
          "_$AN": undefined,
          "element": <img
            alt="image"
            class="max-w-full h-auto transition-shadow ease-in-out duration-300
     
     "
            height=""
            max-height=""
            max-width=""
            min-height=""
            min-width=""
            part="custom-image"
            src=""
            srcset=""
            style="border:undefined undefined undefined"
            width=""
          />,
          "name": "class",
          "options": Object {
            "host": [Circular],
            "isConnected": true,
            "renderBefore": null,
          },
          "strings": Array [
            "max-w-full h-auto transition-shadow ease-in-out duration-300
     ",
            "
     ",
            "",
          ],
          "type": 1,
        },
        tt {
          "_$AH": undefined,
          "_$AM": [Circular],
          "_$AN": undefined,
          "element": <img
            alt="image"
            class="max-w-full h-auto transition-shadow ease-in-out duration-300
     
     "
            height=""
            max-height=""
            max-width=""
            min-height=""
            min-width=""
            part="custom-image"
            src=""
            srcset=""
            style="border:undefined undefined undefined"
            width=""
          />,
          "name": "src",
          "options": Object {
            "host": [Circular],
            "isConnected": true,
            "renderBefore": null,
          },
          "type": 1,
        },
        tt {
          "_$AH": undefined,
          "_$AM": [Circular],
          "_$AN": undefined,
          "element": <img
            alt="image"
            class="max-w-full h-auto transition-shadow ease-in-out duration-300
     
     "
            height=""
            max-height=""
            max-width=""
            min-height=""
            min-width=""
            part="custom-image"
            src=""
            srcset=""
            style="border:undefined undefined undefined"
            width=""
          />,
          "name": "srcset",
          "options": Object {
            "host": [Circular],
            "isConnected": true,
            "renderBefore": null,
          },
          "type": 1,
        },
        tt {
          "_$AH": "image",
          "_$AM": [Circular],
          "_$AN": undefined,
          "element": <img
            alt="image"
            class="max-w-full h-auto transition-shadow ease-in-out duration-300
     
     "
            height=""
            max-height=""
            max-width=""
            min-height=""
            min-width=""
            part="custom-image"
            src=""
            srcset=""
            style="border:undefined undefined undefined"
            width=""
          />,
          "name": "alt",
          "options": Object {
            "host": [Circular],
            "isConnected": true,
            "renderBefore": null,
          },
          "type": 1,
        },
        tt {
          "_$AH": undefined,
          "_$AM": [Circular],
          "_$AN": undefined,
          "element": <img
            alt="image"
            class="max-w-full h-auto transition-shadow ease-in-out duration-300
     
     "
            height=""
            max-height=""
            max-width=""
            min-height=""
            min-width=""
            part="custom-image"
            src=""
            srcset=""
            style="border:undefined undefined undefined"
            width=""
          />,
          "name": "width",
          "options": Object {
            "host": [Circular],
            "isConnected": true,
            "renderBefore": null,
          },
          "type": 1,
        },
        tt {
          "_$AH": undefined,
          "_$AM": [Circular],
          "_$AN": undefined,
          "element": <img
            alt="image"
            class="max-w-full h-auto transition-shadow ease-in-out duration-300
     
     "
            height=""
            max-height=""
            max-width=""
            min-height=""
            min-width=""
            part="custom-image"
            src=""
            srcset=""
            style="border:undefined undefined undefined"
            width=""
          />,
          "name": "height",
          "options": Object {
            "host": [Circular],
            "isConnected": true,
            "renderBefore": null,
          },
          "type": 1,
        },
        tt {
          "_$AH": undefined,
          "_$AM": [Circular],
          "_$AN": undefined,
          "element": <img
            alt="image"
            class="max-w-full h-auto transition-shadow ease-in-out duration-300
     
     "
            height=""
            max-height=""
            max-width=""
            min-height=""
            min-width=""
            part="custom-image"
            src=""
            srcset=""
            style="border:undefined undefined undefined"
            width=""
          />,
          "name": "min-width",
          "options": Object {
            "host": [Circular],
            "isConnected": true,
            "renderBefore": null,
          },
          "type": 1,
        },
        tt {
          "_$AH": undefined,
          "_$AM": [Circular],
          "_$AN": undefined,
          "element": <img
            alt="image"
            class="max-w-full h-auto transition-shadow ease-in-out duration-300
     
     "
            height=""
            max-height=""
            max-width=""
            min-height=""
            min-width=""
            part="custom-image"
            src=""
            srcset=""
            style="border:undefined undefined undefined"
            width=""
          />,
          "name": "max-width",
          "options": Object {
            "host": [Circular],
            "isConnected": true,
            "renderBefore": null,
          },
          "type": 1,
        },
        tt {
          "_$AH": undefined,
          "_$AM": [Circular],
          "_$AN": undefined,
          "element": <img
            alt="image"
            class="max-w-full h-auto transition-shadow ease-in-out duration-300
     
     "
            height=""
            max-height=""
            max-width=""
            min-height=""
            min-width=""
            part="custom-image"
            src=""
            srcset=""
            style="border:undefined undefined undefined"
            width=""
          />,
          "name": "min-height",
          "options": Object {
            "host": [Circular],
            "isConnected": true,
            "renderBefore": null,
          },
          "type": 1,
        },
        tt {
          "_$AH": undefined,
          "_$AM": [Circular],
          "_$AN": undefined,
          "element": <img
            alt="image"
            class="max-w-full h-auto transition-shadow ease-in-out duration-300
     
     "
            height=""
            max-height=""
            max-width=""
            min-height=""
            min-width=""
            part="custom-image"
            src=""
            srcset=""
            style="border:undefined undefined undefined"
            width=""
          />,
          "name": "max-height",
          "options": Object {
            "host": [Circular],
            "isConnected": true,
            "renderBefore": null,
          },
          "type": 1,
        },
        tt {
          "_$AH": Array [
            "undefined undefined undefined",
          ],
          "_$AM": [Circular],
          "_$AN": undefined,
          "element": <img
            alt="image"
            class="max-w-full h-auto transition-shadow ease-in-out duration-300
     
     "
            height=""
            max-height=""
            max-width=""
            min-height=""
            min-width=""
            part="custom-image"
            src=""
            srcset=""
            style="border:undefined undefined undefined"
            width=""
          />,
          "name": "style",
          "options": Object {
            "host": [Circular],
            "isConnected": true,
            "renderBefore": null,
          },
          "strings": Array [
            "border:",
            "",
          ],
          "type": 1,
        },
      ],
    },
    "_$AM": undefined,
    "_$AN": undefined,
    "_$Cm": true,
    "options": Object {
      "host": [Circular],
      "isConnected": true,
      "renderBefore": null,
    },
    "type": 2,
  },
  "_$E_": Promise {},
  "_$Ei": undefined,
  "_$El": null,
  "enableUpdating": [Function],
  "hasUpdated": true,
  "isUpdatePending": false,
  "renderOptions": Object {
    "host": [Circular],
    "isConnected": true,
    "renderBefore": null,
  },
  "renderRoot": ShadowRoot {
    "_$litPart$": W {
      "_$AA": <!---->,
      "_$AB": null,
      "_$AH": Ee {
        "_$AD": B {
          "el": <template />,
          "parts": Array [
            Object {
              "ctor": [Function],
              "index": 0,
              "name": "class",
              "strings": Array [
                "max-w-full h-auto transition-shadow ease-in-out duration-300
     ",
                "
     ",
                "",
              ],
              "type": 1,
            },
            Object {
              "ctor": [Function],
              "index": 0,
              "name": "src",
              "strings": Array [
                "",
                "",
              ],
              "type": 1,
            },
            Object {
              "ctor": [Function],
              "index": 0,
              "name": "srcset",
              "strings": Array [
                "",
                "",
              ],
              "type": 1,
            },
            Object {
              "ctor": [Function],
              "index": 0,
              "name": "alt",
              "strings": Array [
                "",
                "",
              ],
              "type": 1,
            },
            Object {
              "ctor": [Function],
              "index": 0,
              "name": "width",
              "strings": Array [
                "",
                "",
              ],
              "type": 1,
            },
            Object {
              "ctor": [Function],
              "index": 0,
              "name": "height",
              "strings": Array [
                "",
                "",
              ],
              "type": 1,
            },
            Object {
              "ctor": [Function],
              "index": 0,
              "name": "min-width",
              "strings": Array [
                "",
                "",
              ],
              "type": 1,
            },
            Object {
              "ctor": [Function],
              "index": 0,
              "name": "max-width",
              "strings": Array [
                "",
                "",
              ],
              "type": 1,
            },
            Object {
              "ctor": [Function],
              "index": 0,
              "name": "min-height",
              "strings": Array [
                "",
                "",
              ],
              "type": 1,
            },
            Object {
              "ctor": [Function],
              "index": 0,
              "name": "max-height",
              "strings": Array [
                "",
                "",
              ],
              "type": 1,
            },
            Object {
              "ctor": [Function],
              "index": 0,
              "name": "style",
              "strings": Array [
                "border:",
                "",
              ],
              "type": 1,
            },
          ],
        },
        "_$AM": [Circular],
        "_$AN": undefined,
        "u": Array [
          tt {
            "_$AH": Array [
              undefined,
              undefined,
            ],
            "_$AM": [Circular],
            "_$AN": undefined,
            "element": <img
              alt="image"
              class="max-w-full h-auto transition-shadow ease-in-out duration-300
     
     "
              height=""
              max-height=""
              max-width=""
              min-height=""
              min-width=""
              part="custom-image"
              src=""
              srcset=""
              style="border:undefined undefined undefined"
              width=""
            />,
            "name": "class",
            "options": Object {
              "host": [Circular],
              "isConnected": true,
              "renderBefore": null,
            },
            "strings": Array [
              "max-w-full h-auto transition-shadow ease-in-out duration-300
     ",
              "
     ",
              "",
            ],
            "type": 1,
          },
          tt {
            "_$AH": undefined,
            "_$AM": [Circular],
            "_$AN": undefined,
            "element": <img
              alt="image"
              class="max-w-full h-auto transition-shadow ease-in-out duration-300
     
     "
              height=""
              max-height=""
              max-width=""
              min-height=""
              min-width=""
              part="custom-image"
              src=""
              srcset=""
              style="border:undefined undefined undefined"
              width=""
            />,
            "name": "src",
            "options": Object {
              "host": [Circular],
              "isConnected": true,
              "renderBefore": null,
            },
            "type": 1,
          },
          tt {
            "_$AH": undefined,
            "_$AM": [Circular],
            "_$AN": undefined,
            "element": <img
              alt="image"
              class="max-w-full h-auto transition-shadow ease-in-out duration-300
     
     "
              height=""
              max-height=""
              max-width=""
              min-height=""
              min-width=""
              part="custom-image"
              src=""
              srcset=""
              style="border:undefined undefined undefined"
              width=""
            />,
            "name": "srcset",
            "options": Object {
              "host": [Circular],
              "isConnected": true,
              "renderBefore": null,
            },
            "type": 1,
          },
          tt {
            "_$AH": "image",
            "_$AM": [Circular],
            "_$AN": undefined,
            "element": <img
              alt="image"
              class="max-w-full h-auto transition-shadow ease-in-out duration-300
     
     "
              height=""
              max-height=""
              max-width=""
              min-height=""
              min-width=""
              part="custom-image"
              src=""
              srcset=""
              style="border:undefined undefined undefined"
              width=""
            />,
            "name": "alt",
            "options": Object {
              "host": [Circular],
              "isConnected": true,
              "renderBefore": null,
            },
            "type": 1,
          },
          tt {
            "_$AH": undefined,
            "_$AM": [Circular],
            "_$AN": undefined,
            "element": <img
              alt="image"
              class="max-w-full h-auto transition-shadow ease-in-out duration-300
     
     "
              height=""
              max-height=""
              max-width=""
              min-height=""
              min-width=""
              part="custom-image"
              src=""
              srcset=""
              style="border:undefined undefined undefined"
              width=""
            />,
            "name": "width",
            "options": Object {
              "host": [Circular],
              "isConnected": true,
              "renderBefore": null,
            },
            "type": 1,
          },
          tt {
            "_$AH": undefined,
            "_$AM": [Circular],
            "_$AN": undefined,
            "element": <img
              alt="image"
              class="max-w-full h-auto transition-shadow ease-in-out duration-300
     
     "
              height=""
              max-height=""
              max-width=""
              min-height=""
              min-width=""
              part="custom-image"
              src=""
              srcset=""
              style="border:undefined undefined undefined"
              width=""
            />,
            "name": "height",
            "options": Object {
              "host": [Circular],
              "isConnected": true,
              "renderBefore": null,
            },
            "type": 1,
          },
          tt {
            "_$AH": undefined,
            "_$AM": [Circular],
            "_$AN": undefined,
            "element": <img
              alt="image"
              class="max-w-full h-auto transition-shadow ease-in-out duration-300
     
     "
              height=""
              max-height=""
              max-width=""
              min-height=""
              min-width=""
              part="custom-image"
              src=""
              srcset=""
              style="border:undefined undefined undefined"
              width=""
            />,
            "name": "min-width",
            "options": Object {
              "host": [Circular],
              "isConnected": true,
              "renderBefore": null,
            },
            "type": 1,
          },
          tt {
            "_$AH": undefined,
            "_$AM": [Circular],
            "_$AN": undefined,
            "element": <img
              alt="image"
              class="max-w-full h-auto transition-shadow ease-in-out duration-300
     
     "
              height=""
              max-height=""
              max-width=""
              min-height=""
              min-width=""
              part="custom-image"
              src=""
              srcset=""
              style="border:undefined undefined undefined"
              width=""
            />,
            "name": "max-width",
            "options": Object {
              "host": [Circular],
              "isConnected": true,
              "renderBefore": null,
            },
            "type": 1,
          },
          tt {
            "_$AH": undefined,
            "_$AM": [Circular],
            "_$AN": undefined,
            "element": <img
              alt="image"
              class="max-w-full h-auto transition-shadow ease-in-out duration-300
     
     "
              height=""
              max-height=""
              max-width=""
              min-height=""
              min-width=""
              part="custom-image"
              src=""
              srcset=""
              style="border:undefined undefined undefined"
              width=""
            />,
            "name": "min-height",
            "options": Object {
              "host": [Circular],
              "isConnected": true,
              "renderBefore": null,
            },
            "type": 1,
          },
          tt {
            "_$AH": undefined,
            "_$AM": [Circular],
            "_$AN": undefined,
            "element": <img
              alt="image"
              class="max-w-full h-auto transition-shadow ease-in-out duration-300
     
     "
              height=""
              max-height=""
              max-width=""
              min-height=""
              min-width=""
              part="custom-image"
              src=""
              srcset=""
              style="border:undefined undefined undefined"
              width=""
            />,
            "name": "max-height",
            "options": Object {
              "host": [Circular],
              "isConnected": true,
              "renderBefore": null,
            },
            "type": 1,
          },
          tt {
            "_$AH": Array [
              "undefined undefined undefined",
            ],
            "_$AM": [Circular],
            "_$AN": undefined,
            "element": <img
              alt="image"
              class="max-w-full h-auto transition-shadow ease-in-out duration-300
     
     "
              height=""
              max-height=""
              max-width=""
              min-height=""
              min-width=""
              part="custom-image"
              src=""
              srcset=""
              style="border:undefined undefined undefined"
              width=""
            />,
            "name": "style",
            "options": Object {
              "host": [Circular],
              "isConnected": true,
              "renderBefore": null,
            },
            "strings": Array [
              "border:",
              "",
            ],
            "type": 1,
          },
        ],
      },
      "_$AM": undefined,
      "_$AN": undefined,
      "_$Cm": true,
      "options": Object {
        "host": [Circular],
        "isConnected": true,
        "renderBefore": null,
      },
      "type": 2,
    },
  },
}
`;
